### DISCUSSION TITLE
Implement `git branch --contains ID`

### DISCUSSION THREAD

**Discussion by dschrempf** [magit/magit#4728] [Ideas]
Hello! I am sorry if this is already implemented, but I didn't find anything in the documentation, the GitHub issues nor the discussion forum.

Quite often, I want to check if a commit with a given ID is present in the source code I am currently editing. See, for example, https://stackoverflow.com/questions/43535132/given-a-commit-id-how-to-determine-if-current-branch-contains-the-commit.

One way to achieve this is using `git branch --contains ID`. Do you think it is feasible to implement this feature, if it is not available yet?

Thanks for your help!

**Comment by kyleam** [comment]
The [`magit-show-refs` transient](https://magit.vc/manual/magit/References-Buffer.html#References-Buffer) has `--contains` argument.  If you're already in the refs buffer, you can press `y` to see the transient.   Otherwise, you can do `C-u y` to see the transient rather than jumping straight into the refs buffer.

https://github.com/magit/magit/blob/80cdadb08efcf3287f6af138608d2a71bd8d0a40/lisp/magit-refs.el#L368-L373

**Comment by dschrempf** [comment]
Thank you. This works.

However, it is a bit of a contrived way to check if the repository in the current state (the currently edited code) contains a commit. I didn't even know what [Git References](https://git-scm.com/book/en/v2/Git-Internals-Git-References) were.

**Reply by kyleam** [reply]
> it is a bit of a contrived way to check if the repository in the current state

Hmm, the Magit refs buffer is what covers the functionality of `git branch [--list]` (and more generally listing refs), so I don't quite see how it's more contrived than `git branch --contains ID`.

It's not something I have any plans to polish and propose for inclusion in Magit, but perhaps you'd find this command from my personal config more direct:

```elisp
(defun km/magit-rev-ancestor-p (rev-a rev-b)
  "Report whether REV-A is an ancestor of REV-B."
  (interactive
   (let ((rev-a (magit-read-branch-or-commit "Ancestor candidate")))
     (list rev-a (magit-read-other-branch-or-commit
                  (format "Is %s an ancestor of" rev-a)
                  rev-a))))
  (message "%s is %san ancestor of %s" rev-a
           (if (magit-git-success "merge-base" "--is-ancestor"
                                  rev-a rev-b)
               "" "NOT ")
           rev-b))
```

If you only care about "is X in HEAD?", you could modify that to drop `rev-b` and hard code `HEAD`.

**Reply by dschrempf** [reply]
Thanks for the snippet.

The problem for me was that I was not completely familiar with the term "ref". For example, I didn't know that branches are also references, so I would never have searched this functionality under the Magit refs buffer. Further, I was only interested if the commit was in HEAD, as you correctly state. Filtering all refs containing HEAD is also answering this question, but may be a bit of an overkill.
